src/engine.rs

```diff
@@ -647,53 +702,22 @@ impl BindMap {
     }
 }
 
-pub struct DownloadsMapped<'a>(
-    HashMap<
-        Id,
-        (
-            BufferSlice<'a>,
-            GenericOneshotReceiver<RawMutex, Result<(), BufferAsyncError>>,
-        ),
-    >,
-);
-
-impl Downloads {
-    // Discussion: should API change so we get one buffer, rather than mapping all?
-    pub fn map(&self) -> DownloadsMapped {
-        let mut map = HashMap::new();
-        for (id, buf) in &self.buf_map {
-            let buf_slice = buf.slice(..);
-            let (sender, receiver) = futures_intrusive::channel::shared::oneshot_channel();
-            buf_slice.map_async(wgpu::MapMode::Read, move |v| sender.send(v).unwrap());
-            map.insert(*id, (buf_slice, receiver));
-        }
-        DownloadsMapped(map)
-    }
-}
-
-impl<'a> DownloadsMapped<'a> {
-    pub async fn get_mapped(&self, proxy: BufProxy) -> Result<BufferView, Error> {
-        let (slice, recv) = self.0.get(&proxy.id).ok_or("buffer not in map")?;
-        if let Some(recv_result) = recv.receive().await {
-            recv_result?;
-        } else {
-            return Err("channel was closed".into());
-        }
-        Ok(slice.get_mapped_range())
-    }
-}
-
 const SIZE_CLASS_BITS: u32 = 1;
 
 impl ResourcePool {
     /// Get a buffer from the pool or create one.
-    fn get_buf(&mut self, proxy: &BufProxy, usage: BufferUsages, device: &Device) -> Buffer {
-        let rounded_size = Self::size_class(proxy.size, SIZE_CLASS_BITS);
+    fn get_buf(
+        &mut self,
+        size: u64,
+        name: &'static str,
+        usage: BufferUsages,
+        device: &Device,
+    ) -> Buffer {
+        let rounded_size = Self::size_class(size, SIZE_CLASS_BITS);
         let props = BufferProperties {
             size: rounded_size,
             usages: usage,
-            #[cfg(feature = "buffer_labels")]
-            name: proxy.name,
+            name: name,
```

Comment:
This change is just a refactor, doesn't change semantics. My read is that the current state of main makes the pool per-name. That is necessary to avoid buffers being misnamed. A global pool would be more efficient, but I don't think it makes a huge difference in practice.

Reply:
So, I intentionally left it as a global pool *unless* the `buffer_labels` feature is enabled, which isn't needed except for debugging (which is why it's enabled for the winit example, but not by default).
This makes it unconditionally a pool per name.

If there's no need to require it to be a global pool, then that's fine. But we should remove the feature entirely in that case

Reply:
Ah, got it. The current state doesn't compile with the feature off. I'll fix that now. I was making changes and was relying on rust-analyzer to tell me what was wrong, but it only looks at one set of features.